### ISSUE TITLE
Map typing or the lack thereof

### ISSUE THREAD

**Issue by rtomsick** [swagger-api/swagger-core#2356]
Currently, there appears to be a rather large deficiency in the way that maps are handled, affecting a relatively common pattern.

Consider the following usage code:

```java

@JsonGetter("some-property")
@ApiModelProperty(value = "some example property")
public Map<MyEnum, Boolean>	
someProp()
{ /*...*/ }

@ApiModel
public static enum MyEnum
{
	@JsonProperty("foo")
	FOO,
	@JsonProperty("bar")
	BAR;
}

```

Now one would expect the resulting schema to reflect that there exists a map whose keys are in the specified enumeration enumeration.  (Note: I understand why for Maps with an object as their key this isn't possible, but for Maps with enums as keys it logically should be possible since there is already a clear enum value -> string serialization strategy).

In reality, we get the following schema

```json
"some-property" : {
  "type" : "object",
  "description" : "some example property",
  "additionalProperties" : {
    "type" : "boolean"
}
```

Which means that any examples generated when displaying the documentation will give the impression that the property names may be any arbitrary string, which is naturally quite wrong.

OK, so what about if we specify a type manually?  Doesn't work either.

Case in point:

```java
@JsonGetter("some-property")
@ApiModelProperty(value = "some example property",
		dataType = "Map[com.example.Example$MyEnum,boolean]")
public Map<MyEnum, Boolean>	
someProp()
{ /*...*/ }


```
Same result.  (Same for inner classes/enums as those in their own file, BTW.)

Indeed, it appears that the only way to get the desired behavior (sane examples in the resulting documentation, a strict schema of what may be used as the "key", aka. property name(s) of the object) is to create a dummy object which duplicates the enum entries string representations, like so:

```java
private abstract class DummyExample
{
	@JsonProperty("foo")
	public boolean foo;

	@JsonProperty("bar")
	public boolean bar;
}
```

Then use the dummy object via `@ApiModelProperty#dataType`, which is kinda... hackish.  It works and produces exactly the right result spec-wise, but it's a pain in the ass maintenance-wise.  (To be fair though, until #2352 is fixed this is *still* a necessary workaround to get the right enum values in the spec...)

It would therefore like to propose that swagger-core address this use case (enum as Map key) explicitly, as it's both a fairly common and a fairly useful pattern.

Thanks!

**Comment by webron** [comment]
There's nothing we can do about it as the spec doesn't allow defining key values in a map. To do that, we'd need support for `patternProperties` which is not part of the spec.

**Comment by rtomsick** [comment]
I understand that, and that's why I wouldn't propose a change in behavior for `Map<K, V>` where `K` is an arbitrary class.

But I would contend that since `Map<MyEnum, Boolean>` is logically equivalent to `DummyExample` in the code above, there **is** something that can be done about the cases where `K` is an enum.

As the set of possible keys is both finite and known at compile time, it would be logically-sound to produce a representation of the `Map` as though it were simply an object with those properties (all optional).

**Comment by webron** [comment]
Sure, I see what you're saying. The only thing is that this would be a one-way conversion. Converting API definitions like that to code would not result an an enum map. A PR to support this change would be welcome.

**Comment by rtomsick** [comment]
That's true, there's potentially no way to go from OpenAPI spec -> code, but that's the same problem with enums in general isn't it?  

I don't use client codegen or anything, so I have to plead ignorance here.  How are enums currently handled?

**Comment by PeaceInColour** [comment]
I have a fix for the original issue here with `@JsonProperty` not being used for enum names (the issue is still in 1.5.18). The problems is that ModelResolver is using a deprecated method to read enum values. The deprecated method doesn't check for any override values and just defaults to the enum names in the code. Changing the deprecated method to the replacement method allows for `@JsonProperty` to rename enums.

**Comment by arathai** [comment]
Hi,

I had the same issue but was using EnumMap instead of using plain map so there's a pretty strong guarantee that only the known Enum instances are valid keys. Therefore I wrote this ModelConverter.

Beware that I use Enum.name() and I'm currently not checking if for example SerializationFeature.WRITE_ENUMS_USING_TO_STRING is used in the config.

Hope it helps!

```
import java.util.Collections;
import java.util.Iterator;
import java.util.EnumMap;

import com.fasterxml.jackson.databind.type.MapType;

import io.swagger.v3.core.converter.AnnotatedType;
import io.swagger.v3.core.converter.ModelConverter;
import io.swagger.v3.core.converter.ModelConverterContext;
import io.swagger.v3.core.jackson.TypeNameResolver;
import io.swagger.v3.oas.models.media.Schema;


/**
 * @author <a href="mailto:Alexander.Rathai@Materna.de">Alexander Rathai</a>
 */
public class EnumMapModelConverter implements ModelConverter {

	/**
	 * @see io.swagger.v3.core.converter.ModelConverter#resolve(io.swagger.v3.core.converter.AnnotatedType, io.swagger.v3.core.converter.ModelConverterContext,
	 *      java.util.Iterator)
	 */
	@Override
	public Schema resolve(AnnotatedType type, ModelConverterContext context, Iterator<ModelConverter> chain) {
		java.lang.reflect.Type reflectiveType = type.getType();
		if( reflectiveType instanceof MapType ) {
			MapType mapType = (MapType)reflectiveType;
			Class<?> rawClass = mapType.getRawClass();
			if( EnumMap.class.isAssignableFrom(rawClass) ) {
				String nameForType = TypeNameResolver.std.nameForType(mapType, Collections.emptySet());
				String nameForContentType = TypeNameResolver.std
						.nameForType(mapType.getContentType(), Collections.emptySet());
				Schema enumMapSchema = new Schema<>().type("object").name(nameForType);
				AnnotatedType annotatedContentType = new AnnotatedType(mapType.getContentType())
						.parent(enumMapSchema)
						.name(nameForContentType)
						.resolveAsRef(true);
				Schema enumMapContentSchema = chain.next().resolve(annotatedContentType, context, chain);
				/*
				 * this cast is considered safe because we checked for EnumMap earlier
				 */
				@SuppressWarnings("unchecked")
				Class<? extends Enum> enumClass = (Class<? extends Enum>)mapType.getKeyType().getRawClass();
				Enum[] enumConstants = enumClass.getEnumConstants();
				for( Enum object : enumConstants ) {
					enumMapSchema.addProperties(object.name(), enumMapContentSchema);
				}
				return enumMapSchema;
			}
		}
		if( chain.hasNext() ) {
			return chain.next().resolve(type, context, chain);
		}
		return null;
	}
}
```
